### DISCUSSION TITLE
Typescript error when using Apollo client to make SSR requests with NEXTJS

### DISCUSSION THREAD

**Discussion by captDaylight** [dotansimha/graphql-code-generator#8282] [Q&A] [answered]
Using the plugin `typescript-react-apollo` has generally been a breeze. However, in some scenarios I can't use the generated hooks and need to use NextJS's `getStaticProps` function to fetch data at build time. 

For example, imagine I want to grab products at build. In `getStaticProps` I won't be using the generated hook `useGetProductsQuery`, rather the query document and request result. I can grab these types from codegen's generated types: `GetProductsQueryResult` and `GetProductsDocument`. 

```tsx
export const getStaticProps = async () => {
  const client = ApolloClient();

  const { data }: GetProductsQueryResult = await client.query({
    query: GetProductsDocument,
    variables: {
      limit: 10,
    },
  });
  // ...
}
```

However, the `const { data } = ...` line is returning a typescript error:

```bash
Type 'ApolloQueryResult' is missing the following properties from type 'QueryResult { limit?: InputMaybe  | undefined; }>>': client, observable, called, variables, and 7 more.
```

I can see the other variables that the error is referencing on [`QueryResult`](https://github.com/apollographql/apollo-client/blob/b8cebdae5eaadba4761abe571c858b3c8207843b/src/react/types/types.ts#L67), however [`ApolloQueryResult`](https://github.com/apollographql/apollo-client/blob/8ec45a995380c6a8e6fa4b4ba3abb4bbfad97c81/src/core/types.ts#L135) seems to be the return type on [`client.query`](https://github.com/apollographql/apollo-client/blob/8ec45a995380c6a8e6fa4b4ba3abb4bbfad97c81/src/core/ApolloClient.ts#L314).

The generated Apollo query looks like:
```ts
export type GetProductsQueryResult = Apollo.QueryResult<GetProductsQuery, GetProductsQueryVariables>;
```

Shouldn't this be `ApolloQueryResult` not `QueryResult`? Am I approaching this wrong?

**Comment by charlypoly** [comment]
Hi @captDaylight,

I would suggest you use a simpler setup that is well integrated with the Apollo client: [the `typed-document-node` plugin](https://www.graphql-code-generator.com/plugins/typescript/typed-document-node).

By using this plugin, you will get typed GraphQL documents (ex: `GetProductsDocument`) that work natively with Apollo and most GraphQL clients.

**Comment by captDaylight** [comment, marked as answer]
Hi @charlypoly, thanks for the advice. However, with my current setup I'm already generating the `GetProductsDocument`. My issue was I was importing the wrong thing from the generated file, instead of `GetProductsQueryResult` it should have been `GetProductsQuery`

So for example this instead:
```ts
// ...
import type { GetProductsQuery } from 'yadayada/generated/graphql';
import { GetProductsDocument } from 'yadayada/generated/graphql';

export const getStaticProps = async () => {
  const client = ApolloClient();

  const { data } = await client.query<GetProductsQuery>({
    query: GetProductsDocument,
    variables: {
      limit: PRODUCTS_PAGE_LIMIT,
    },
  });
  // ...
}
```
Now there errors are gone 🎉
